### PR TITLE
Fix crashes in shadow server

### PR DESCRIPTION
**Pull request by xhaakon**

### CODE REVIEW

**Comment by hardening** [comment]
@freerdp-bot would you mind test this PR and see if it compiles cleanly, thank you.

**Comment by hardening** [comment]
It looks good to me, +1.
@xhaakon it would be really great, if you could write a unitary test (complementing an existing one or creating a new one) that would verify your correction.

File: winpr/libwinpr/synch/barrier.c [hunk 1]
```diff
@@ -165,6 +167,8 @@ BOOL WINAPI DeleteSynchronizationBarrier(LPSYNCHRONIZATION_BARRIER lpBarrier)
 	if (!pBarrier)
 		return TRUE;
 
+	while (InterlockedCompareExchange(&pBarrier->count, 0, 0) != 0);
+
```
Comment (bmiklautz) [inline winpr/libwinpr/synch/barrier.c]: For a high count this might cause the CPU to get busy. Maybe a minimal Sleep should be added here.

**Comment by bmiklautz** [comment]
As hardening suggested. If you could add a unit test that reproduces this problem that would be great.

**Comment by realjiangms** [comment]
This also fixes the race condition I pasted at https://github.com/FreeRDP/FreeRDP/pull/2395#issuecomment-77742784

File: winpr/libwinpr/synch/barrier.c (hunk 1 above)
Comment (xhaakon) [inline winpr/libwinpr/synch/barrier.c]: I've added 100ms sleep into the loop.

**Comment by xhaakon** [comment]
@bmiklautz I've extended TestSynchBarrier to cover also the case of barrier's early deletion. Please let me know if you'd rather prefer a test in some different format.

**Comment by bmiklautz** [comment]
@freerdp-bot test

**Pull request merged**